File: warehouse-api/src/password-reset/password-reset.service.ts
import { Injectable, NotFoundException, BadRequestException } from '@nestjs/common';
import { InjectModel } from '@nestjs/mongoose';
import { Model } from 'mongoose';
import * as bcrypt from 'bcrypt';
import { PasswordResetRequest, PasswordResetRequestDocument } from '../schemas/password-reset-request.schema';
import { User, UserDocument } from '../schemas/user.schema';
import { CreateResetRequestDto, ResolveResetRequestDto } from './dto/password-reset.dto';

@Injectable()
export class PasswordResetService {
  constructor(
    @InjectModel(PasswordResetRequest.name)
    private resetModel: Model<PasswordResetRequestDocument>,
    @InjectModel(User.name)
    private userModel: Model<UserDocument>,
  ) {}

  /** Public: user submits a reset request (no auth required) */
  async createRequest(dto: CreateResetRequestDto) {
    const user = await this.userModel.findOne({ username: dto.username.toLowerCase() });
    if (!user) throw new NotFoundException('用户名不存在，请确认后重试');

    // Allow one pending request per user at a time
    const existing = await this.resetModel.findOne({
      username: dto.username.toLowerCase(),
      status: 'pending',
    });
    if (existing) {
      return { message: '已有待处理的申请，请等待管理员处理' };
    }

    await this.resetModel.create({
      username: user.username,
      displayName: user.name,
      userNote: dto.userNote ?? '',
      status: 'pending',
    });

    return { message: '申请已提交，请联系管理员处理' };
  }

  /** Admin: list all requests */
  async findAll(status?: string) {
    const filter: any = {};
    if (status) filter.status = status;
    return this.resetModel.find(filter).sort({ createdAt: -1 }).lean();
  }

  /** Admin: resolve a request (complete or reject) */
  async resolve(id: string, dto: ResolveResetRequestDto, adminName: string) {
    const req = await this.resetModel.findById(id);
    if (!req) throw new NotFoundException('申请记录不存在');
    if (req.status !== 'pending') {
      throw new BadRequestException('该申请已处理，无法重复操作');
    }

    if (dto.status === 'completed') {
      if (!dto.newPassword || dto.newPassword.length < 6) {
        throw new BadRequestException('重置密码不能少于 6 位');
      }
      const user = await this.userModel.findOne({ username: req.username });
      if (!user) throw new NotFoundException('对应用户不存在');

      user.passwordHash = await bcrypt.hash(dto.newPassword, 10);
      user.mustChangePassword = true;
      await user.save();
    }

    req.status = dto.status;
    req.adminNote = dto.adminNote ?? '';
    req.resolvedAt = new Date();
    req.resolvedBy = adminName;
    await req.save();

    return { message: dto.status === 'completed' ? '密码已重置，用户下次登录时必须修改密码' : '申请已拒绝' };
  }

  /** Admin: delete a request record */
  async remove(id: string) {
    await this.resetModel.findByIdAndDelete(id);
    return { message: '记录已删除' };
  }
}

File: flutter_warehouse/lib/services/password_reset_service.dart
import 'api_service.dart';

class PasswordResetRequest {
  final String id;
  final String username;
  final String displayName;
  final String status; // pending | completed | rejected
  final String adminNote;
  final String userNote;
  final DateTime createdAt;
  final DateTime? resolvedAt;
  final String resolvedBy;

  PasswordResetRequest({
    required this.id,
    required this.username,
    required this.displayName,
    required this.status,
    required this.adminNote,
    required this.userNote,
    required this.createdAt,
    this.resolvedAt,
    required this.resolvedBy,
  });

  factory PasswordResetRequest.fromJson(Map<String, dynamic> json) =>
      PasswordResetRequest(
        id: json['_id'] ?? '',
        username: json['username'] ?? '',
        displayName: json['displayName'] ?? '',
        status: json['status'] ?? 'pending',
        adminNote: json['adminNote'] ?? '',
        userNote: json['userNote'] ?? '',
        createdAt: json['createdAt'] != null
            ? DateTime.parse(json['createdAt']).toLocal()
            : DateTime.now(),
        resolvedAt: json['resolvedAt'] != null
            ? DateTime.parse(json['resolvedAt']).toLocal()
            : null,
        resolvedBy: json['resolvedBy'] ?? '',
      );

  String get statusLabel => switch (status) {
        'pending' => '待处理',
        'completed' => '已完成',
        'rejected' => '已拒绝',
        _ => status,
      };
}

class PasswordResetService {
  final _api = ApiService.instance.dio;

  /// Public — submit a reset request (no login required)
  Future<String> submitRequest({
    required String username,
    String? userNote,
  }) async {
    final response = await _api.post('/password-reset/request', data: {
      'username': username,
      if (userNote != null && userNote.isNotEmpty) 'userNote': userNote,
    });
    return response.data['message'] as String;
  }

  /// Admin — list all requests
  Future<List<PasswordResetRequest>> getAll({String? status}) async {
    final response = await _api.get('/password-reset', queryParameters: {
      if (status != null) 'status': status,
    });
    return (response.data as List)
        .map((e) => PasswordResetRequest.fromJson(e as Map<String, dynamic>))
        .toList();
  }

  /// Admin — resolve a request
  Future<String> resolve({
    required String id,
    required String status, // 'completed' | 'rejected'
    String? adminNote,
    String? newPassword,
  }) async {
    final response = await _api.patch('/password-reset/$id/resolve', data: {
      'status': status,
      if (adminNote != null && adminNote.isNotEmpty) 'adminNote': adminNote,
      if (newPassword != null && newPassword.isNotEmpty) 'newPassword': newPassword,
    });
    return response.data['message'] as String;
  }

  /// Admin — delete a request record
  Future<void> remove(String id) async {
    await _api.delete('/password-reset/$id');
  }
}
